plugin/evm/atomic/txpool/txs.go

```diff
@@ -194,51 +206,53 @@ func (t *Txs) CancelCurrentTx(txID ids.ID) {
 	}
 }
 
-// [CancelCurrentTxs] marks the attempt to issue [currentTxs]
-// as being aborted. If this is called after a buildBlock error
-// caused by the atomic transaction, then DiscardCurrentTx should have been called
-// such that this call will have no effect and should not re-issue the invalid tx.
+// CancelCurrentTxs attempts to mark all Current transactions as Pending.
+//
+// This should be called after building a block failed due to an error unrelated
+// to the transactions.
 func (t *Txs) CancelCurrentTxs() {
 	t.lock.Lock()
 	defer t.lock.Unlock()
 
-	// If building a block failed, put the currentTx back in [txs]
-	// if it exists.
 	for _, tx := range t.currentTxs {
 		t.cancelTx(tx)
 	}
 
-	// If there are more transactions to be issued, add an item
-	// to Pending.
-	if t.txHeap.Len() > 0 {
+	// Since this function is called after block building has failed, we should
+	// make sure to signal if we are willing to build another block.
+	if t.pendingTxs.Len() > 0 {
 		t.addPending()
 	}
 }
 
-// cancelTx removes [tx] from current transactions and moves it back into the
-// tx heap.
-// assumes the lock is held.
+// cancelTx attempts to mark the Current transaction as Pending. If the tx can
+// not be marked as Pending, it will be Discarded.
+//
+// Assumes the lock is held.
 func (t *Txs) cancelTx(tx *atomic.Tx) {
-	// Add tx to heap sorted by gasPrice
+	txID := tx.ID()
 	gasPrice, err := t.atomicTxGasPrice(tx)
-	if err == nil {
-		t.txHeap.Push(tx, gasPrice)
-		t.metrics.pendingTxs.Update(int64(t.txHeap.Len()))
-	} else {
-		// If the err is not nil, we simply discard the transaction because it is
-		// invalid. This should never happen but we guard against the case it does.
-		log.Error("failed to calculate atomic tx gas price while canceling current tx", "err", err)
-		t.removeSpenders(tx)
-		t.discardedTxs.Put(tx.ID(), tx)
-		t.metrics.discardedTxs.Inc(1)
+	if err != nil {
+		log.Error("failed to calculate atomic tx gas price while canceling current tx",
+			"txID", txID,
+			"err", err,
+		)
+		t.discardCurrentTx(tx)
```

Comment:
Previously we had copied the logic from `discardCurrentTx` to here, now we just call the function.